Update Airflow AWS MWAA deployment docs (#3860)
kedro-org/kedro: docs/source/deployment/airflow.md

![](../meta/images/astronomer_cloud_deployment.png)

## How to run a Kedro pipeline on Amazon AWS Managed Workflows for Apache Airflow (MWAA)

### Kedro project preparation

> MWAA, or Managed Workflows for Apache Airflow, is an AWS service that makes it easier to set up, operate, and scale Apache Airflow in the cloud. Deploying a Kedro pipeline to MWAA is similar to Astronomer, but there are some key differences: you need to store your project data in an AWS S3 bucket and make necessary changes to your Data Catalog. Additionally, you must configure how you upload your Kedro configuration, install your Kedro package, and set up the necessary environment variables.

Reviewer: ```suggestion
MWAA, or Managed Workflows for Apache Airflow, is an AWS service that makes it easier to set up, operate, and scale Apache Airflow in the cloud. Deploying a Kedro pipeline to MWAA is similar to Astronomer, but there are some key differences: you need to store your project data in an AWS S3 bucket and make necessary changes to your `DataCatalog`. Additionally, you must configure how you upload your Kedro configuration, install your Kedro package, and set up the necessary environment variables.
```

1. Complete steps 1-4 from the [Create, prepare and package example Kedro project](#create-prepare-and-package-example-kedro-project) section.
2. Your project's data should not reside in the working directory of the Airflow container. Instead, [create an S3 bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/creating-bucket.html) and [upload your data folder from the new-kedro-project folder to your S3 bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/upload-objects.html).

> 3. Modify the Data Catalog to reference data in your S3 bucket by updating the filepath and add credentials line for each Dataset in the `new-kedro-project/conf/airflow/catalog.yml`. Add the S3 prefix to the filepath as shown below:

Reviewer: ```suggestion
3. Modify the `DataCatalog` to reference data in your S3 bucket by updating the filepath and add credentials line for each dataset in `new-kedro-project/conf/airflow/catalog.yml`. Add the S3 prefix to the filepath as shown below:
```

```shell
companies:
  type: pandas.CSVDataset

>   filepath: s3://your_S3_bucket/data/01_raw/companies.csv

Reviewer 2: ```suggestion
  filepath: s3://<your_S3_bucket>/data/01_raw/companies.csv
```
To indicate that this needs to be filled in by the user

  credentials: dev_s3
```
4. [Set up AWS credentials](https://docs.aws.amazon.com/keyspaces/latest/devguide/access.credentials.html) to provide read and write access to your S3 bucket. Update `new-kedro-project/conf/local/credentials.yml` with your AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY:
```shell
dev_s3:
  client_kwargs:
    aws_access_key_id: *********************
    aws_secret_access_key: ******************************************

> ```

Reviewer 2: I haven't tried this yet, but do the credentials in the `local/credentials.yml` work on the cloud platform? This file ideally shouldn't be uploaded to the cloud at all. Maybe setting up env variables?

Author: I tried that approach with the `local` environment, and it works. We can change the environment from `local` to `prod`, but am I correct in understanding that there isn't a big difference? 

For setting environment variables, I am using an `env_var.sh` file that is located in the same S3 bucket as the `conf` folder, which is currently packed into `plugins.zip`. Therefore, I don't think it makes much sense to use environment variables for security reasons because, ultimately, they will be stored in the same S3 bucket. However, this S3 bucket is secure and does not have public access, so I think it's not a problem.

> 5. Add `s3fs` to your project’s `requirements.txt` in `new-kedro-project` to facilitate communication with AWS S3. Some libraries could cause dependency conflicts in the Airflow environment, so make sure to minimise the list and avoid using `kedro-viz` and `pytest`.

Reviewer: ```suggestion
5. Add `s3fs` to your project’s `requirements.txt` in `new-kedro-project` to facilitate communication with AWS S3. Some libraries could cause dependency conflicts in the Airflow environment, so make sure to minimise the requirements list and avoid using `kedro-viz` and `pytest`.
```

```shell
s3fs
```

> 6. Archive your `conf` folder into a `conf.zip` file and upload it to `s3://your_S3_bucket` for later use in the Airflow container. This file will be unzipped into the `plugins` folder within the MWAA Airflow container.

Reviewer 2: Worth noting that the `kedro package` command will create an archived `conf-<package-name>.tar.gz` file in the `dist/` folder without the local conf which should be used here ^

Author: Do you have any suggestions on how this file should be used? Should we simply unpack it to extract the `conf` folder, or should we run `kedro run` directly with the `conf` archive?

Reviewer 2: It doesn't need to be unpacked at all, `OmegaConfigLoader` can handle `.tar.gz` and `.zip` confs

Author: I changed that but I don't know how to be with `export KEDRO_LOGGING_CONFIG="plugins/logging.yml"`, I currently copy `logging.yml` separetely, but maybe you know better solution.

7. Follow steps 5-6 from the [Create, prepare and package example Kedro project](#create-prepare-and-package-example-kedro-project) section to package your Kedro project and generate an Airflow DAG.

> 8. Modify the DAG file `new_kedro_project_dag.py` located in the `dags/` folder by adding `, conf_source="plugins/conf"` to the Kedro session creation in the Kedro operator execution function. This change is necessary because your Kedro configuration folder will be stored in the `plugins/conf` folder, not the root directory:

Reviewer 2: ```suggestion
8. Update the DAG file `new_kedro_project_dag.py` located in the `dags/` folder by adding `conf_source="plugins/conf"` to the arguments of `KedroSession.create()` in the Kedro operator execution function. This change is necessary because your Kedro configuration folder will be stored in the `plugins/conf` folder, not the root directory:
```
It looks strange with the comma

    def execute(self, context):
        configure_project(self.package_name)
        with KedroSession.create(project_path=self.project_path,
                                 env=self.env, conf_source="plugins/conf-new_kedro_project.tar.gz") as session:
            session.run(self.pipeline_name, node_names=[self.node_name])
```

### Deployment on AWAA

> 1. Archive your three files: `new_kedro_project-0.1-py3-none-any.wh`l and `conf-new_kedro_project.tar.gz` located in `new-kedro-project/dist`, and `logging.yml` located in `new-kedro-project/conf/` into a file called `plugins.zip` and upload it to `s3://your_S3_bucket`. This archive will be later unpacked to the `/plugins` folder in the working directory of the Airflow container.

Reviewer: ```suggestion
1. Archive your three files: `new_kedro_project-0.1-py3-none-any.whl` and `conf-new_kedro_project.tar.gz` located in `new-kedro-project/dist`, and `logging.yml` located in `new-kedro-project/conf/` into a file called `plugins.zip` and upload it to `s3://your_S3_bucket`. This archive will be later unpacked to the `/plugins` folder in the working directory of the Airflow container.
```

> 1. Archive your three files: `new_kedro_project-0.1-py3-none-any.wh`l and `conf-new_kedro_project.tar.gz` located in `new-kedro-project/dist`, and `logging.yml` located in `new-kedro-project/conf/` into a file called `plugins.zip` and upload it to `s3://your_S3_bucket`. This archive will be later unpacked to the `/plugins` folder in the working directory of the Airflow container.

Reviewer: I think we should provide the command/instructions on how to archive, because not all users might know how to do that.

Author: Good point, I've added it!

> 2. Create a new `requirements.txt` file, add the command to install your Kedro project archived in the previous step, and upload it to `s3://your_S3_bucket`:

Author: Agree, I rephrased that.

```shell
    def execute(self, context):
        configure_project(self.package_name)
        with KedroSession.create(project_path=self.project_path,
                                 env=self.env, conf_source="plugins/conf-new_kedro_project.tar.gz") as session:
            session.run(self.pipeline_name, node_names=[self.node_name])
```

### Deployment on AWAA
1. Archive your three files: `new_kedro_project-0.1-py3-none-any.whl` and `conf-new_kedro_project.tar.gz` located in `new-kedro-project/dist`, and `logging.yml` located in `new-kedro-project/conf/` into a file called `plugins.zip` and upload it to `s3://your_S3_bucket`.
```shell
zip -j plugins.zip dist/new_kedro_project-0.1-py3-none-any.whl dist/conf-new_kedro_project.tar.gz conf/logging.yml
```
This archive will be later unpacked to the `/plugins` folder in the working directory of the Airflow container.

2. Create a new `requirements.txt` file, add the path where your Kedro project will be unpacked in the Airflow container, and upload `requirements.txt` to `s3://your_S3_bucket`:
```shell
./plugins/new_kedro_project-0.1-py3-none-any.whl
```

> Libraries from `requirements.txt` will be installed during container initialization.

Reviewer: ```suggestion
Libraries from `requirements.txt` will be installed during container initialisation.
```
